### ISSUE TITLE
[coqdep] [native] coqdep is not aware of dependencies on native-generated files

### ISSUE THREAD

**Issue by ejgallego** [rocq-prover/rocq#13035]
As of today, `coqdep` is not aware of dependencies on native files. It is not clear if we would like coqdep to handle this, or if the build system should amend `coqdep` output, but in principle, that means that build rules using `coqdep` such as in coq makefile are unsound, for example when building in "sandbox" mode [any non-declared target / dep is not present]

**Comment by gares** [comment]
`coqdep` works in tandem with Coq's Makefile and `coq_makefile`. They don't use "grouped targets" rules `.coq-native/A.cmxs A.vo &: A.v` so there is no point to declare a dependency like `B.vo : A.vo .coq-native/A.cmxs` (also native is optional, and using different make rules in the two cases would be a pain). So the deps are not unsound, they just may not be right for dune.

IMO the right path would be to put the contents of `.coq-native/A.cmxs` into `A.vo` (I see no reason for keeping the file separate) and simplify the file system usage we do during compilation, rather than make build system aware of the mess.

@maximedenes given the recent improvement to the .vo file format, what I propose may be easy (if dynlink can read from a custom channel rather than a real file).

**Comment by maximedenes** [comment]
> given the recent improvement to the .vo file format

I haven't followed that, do you have a pointer?

> if dynlink can read from a custom channel rather than a real file

I don't think it can. https://ocaml.org/releases/4.11/htmlman/libref/Dynlink.html

**Comment by gares** [comment]
https://github.com/coq/coq/pull/12092
in particular the new API let's one write to a segment via an out_channel , I guess one could have `ocamlopt -shared` write in there (at least on unix).

**Comment by ppedrot** [comment]
> I don't think it can.

@maximedenes You can probably hack around this by reimplementing the `Dynlink.loadfile` function.

https://github.com/ocaml/ocaml/blob/f6a5b755f8c05258a4f7fc0f054fb1e656272402/otherlibs/dynlink/dynlink_common.ml#L336

**Comment by ejgallego** [comment]
> They don't use "grouped targets" rules `.coq-native/A.cmxs A.vo &: A.v` so there is no point to declare a dependency like `B.vo : A.vo .coq-native/A.cmxs`

What do you mean by "grouped targets" ? If you mean rules that produce multiple targets, indeed `coqdep` does produce it [correctly]

```
Init/Datatypes.vo Init/Datatypes.glob Init/Datatypes.v.beautified Init/Datatypes.required_vo: Init/Datatypes.v Init/Notations.vo Init/Ltac.vo Init/Logic.vo
```

> (also native is optional, and using different make rules in the two cases would be a pain). So the deps are not unsound, they just may not be right for dune.

If we interpret the output of `coqdep` as a make-formatted dependency graph indeed the rules are unsound, in the sense that some build systems [not only dune] would enforce that only the dependencies are available at build time of an object [for example having a cache]

Dune particularly doesn't care about this as it already adds the extra targets.

**Comment by gares** [comment]
It's not make formatted, it is*for* make.
If its output is not sufficient for dune you can provide a dune mode, but please don't change its output for make

**Comment by ejgallego** [comment]
The output works for dune, again the rules are the ones incorrect specially for make. Example:

- setup 2 files, `A.v` and `B.v`, `B` depends on `A`.
- coqdep will generate `B.vo B.glob: B.v A.vo`
- compile with native, the corresponding cmxs are generated.
- now modify `B.v` and delete `.coq-native/A.cmxs`
- the build fails.

**Comment by gares** [comment]
> delete `.coq-native/A.cmxs`

If the rules of the game is that you can tamper with the build artifacts then you can break anything.
I would not make the "make rules" more complicated to cover this corner case (you are removing an hidden file...).

As I was saying before the rules *and* dependencies would be different in native/non-native mode if you wanted to make the build system aware of .coq-native/ files. Moreover the link between a .vo and the corresponding .cmxs is so tight that they should really be one file.

I know dune purges the build cache between builds, so, if you don't want to change Coq you must make it aware that .cmxs files exist, etc. For Coq in general it would be much better if the coq-native stuff was not a hidden file next to a vo, but a vo segment. And indeed that would make dune support native with no change at all.

**Comment by maximedenes** [comment]
> You can probably hack around this by reimplementing the `Dynlink.loadfile` function.

Ok for the dynlink part. What about the compilation part? How do I pass the `.cmi` files extracted from the `.vo` segment to the OCaml compiler?

Those are not rhetorical questions, if there's a solution for having `native_compute` produce a segment rather than separate files, and interface easily enough with the OCaml compiler, I'm all for it.

**Comment by SkySkimmer** [comment]
Would it be possible / sensible to build the cmxs separately by using the vo? ie instead of current 
~~~gmake
# implicitly builds cmxs too depending on native configuration
%.vo: %.v
    coqc $<

# from coqdep:
B.vo: A.vo # implicitly A.cmxs too depending on native configuration
~~~
we would have 
~~~gmake 
%.vo: %.v
    coqc $< 

%.cmxs: %.vo
    coqnative $<

# from coqdep: 
B.vo: A.vo
B.vo: A.cmxs # depending on native configuration
~~~
AFAIK A.cmxs is not used while producing A.vo

**Comment by gares** [comment]
I totally agree it is not easy, and maybe not feasible in a clean way without patching ocamlopt or using compiler-libs directly.
But here the discussion is about making coqdep & make aware of these compilation details. You just pointed out that B.vo should really depend A.vo .coq-native/a.cmxs *and* .coq-native/a.cmi (and guess that is one wants to have cross module inlining and still be "sound" and .coq-native/a.cmx).

I don't think it is a good idea to make the makefile-based build systems aware of this.

**Comment by gares** [comment]
I don't see how there reules more "sound" (e.g. no .cmi). I don't care if they are unsound, but then I don't know which problem we are trying to fix.

**Comment by maximedenes** [comment]
I suspect it would be much more efficient to setup a meeting with a clear agenda to discuss these issues.

**Comment by ejgallego** [comment]
> Would it be possible / sensible to build the cmxs separately by using the vo? ie instead of current

That would be cool, as we discussed some weeks ago that would allow to have `coq-foo` and `coq-foo-native` opam packages which could prove advantageous [not fully sure that'd be a good idea tho]

**Comment by ppedrot** [comment]
> Those are not rhetorical questions, if there's a solution for having native_compute produce a segment rather than separate files, and interface easily enough with the OCaml compiler, I'm all for it.

Maybe a good start would be to simply rely on a temporary file extracted from the vo? It's probably enough to make it work and the overhead might be negligible compared to the rest of the compilation / linking phase.

**Comment by Blaisorblade** [comment]
> Moreover the link between a .vo and the corresponding .cmxs is so tight that they should really be one file.

What would be the point of having `coqnative` be separate then? As far as I remember, the *motivation* was to enable split packages and do essentially what's proposed in this comment: https://github.com/coq/coq/issues/13035#issuecomment-694881898.
The only difference is that `B.vo: A.cmxs` is not needed for most libraries, you only need `B.cmxs: A.cmxs`. Dependencies from `B.vo` to `A.cmxs` should probably only arise when `B` uses `native_compute` itself.

Otherwise, it seems you should be able to fully build a library to `vo`, install it, and then build the `cmxs` somehow if desired.

**Comment by Blaisorblade** [comment]
*However*, what I propose is a new and nontrivial build system feature, so it's unclear if it should be supported in `coq_makefile`, or if even `coq_makefile` would be the proper place to support it...

**Comment by ejgallego** [comment]
I'm not sure I can follow you, but note that this issue was submitted for the non-split mode, the split mode is a bit different, but the bug still applies as coqdep knows nothing.

**Comment by ppedrot** [comment]
Not having the native files for the local dependencies around is not a real issue, it means that they will be recomputed on the fly by coqc when it encounters a `native_compute` invocation. Obviously this creates an overhead but it does not change the observed behaviour. I think that it's too hard to teach coqdep to recognize that a file requires native precompilation, since it's probably plain undecidable. The most reasonable thing to me is that we should be able to flag specific v files in _CoqProject to act as barriers for compilation of native dependencies. Thus, if you need native compute in a given file you just flag it and the build system will precompile for you.
